zerver/lib/export.py

```diff
@@ -976,28 +984,71 @@ def get_ids(records: List[Record]) -> Set[int]:
         response['zerver_userprofile_mirrordummy'] +
         response['zerver_userprofile_crossrealm'])
 
-    if public_only:
+    consented_user_ids = set()  # type: Set[int]
+    if consent_message_id is not None:
+        consented_user_ids = get_consented_user_ids(consent_message_id)
+
+    if len(consented_user_ids) != 0:
+        public_streams = Stream.objects.filter(realm=realm, invite_only=False)
+        public_stream_recipient_ids = Recipient.objects.filter(
+            type=Recipient.STREAM, type_id__in=public_streams).values_list("id", flat=True)
+
+        consented_recipient_ids = Subscription.objects.filter(user_profile__id__in=consented_user_ids). \
+            values_list("recipient", flat=True)
+
+        recipient_ids = set(public_stream_recipient_ids) | set(consented_recipient_ids)
+        recipient_ids_for_us = get_ids(response['zerver_recipient']) & recipient_ids
+    elif public_only:
         recipient_streams = Stream.objects.filter(realm=realm, invite_only=False)
         recipient_ids = Recipient.objects.filter(
             type=Recipient.STREAM, type_id__in=recipient_streams).values_list("id", flat=True)
         recipient_ids_for_us = get_ids(response['zerver_recipient']) & set(recipient_ids)
     else:
         recipient_ids_for_us = get_ids(response['zerver_recipient'])
 
-    # We capture most messages here, since the
-    # recipients we subscribe to are also the
-    # recipients of most messages we send.
-    messages_we_received = Message.objects.filter(
-        sender__in=ids_of_our_possible_senders,
-        recipient__in=recipient_ids_for_us,
-    ).order_by('id')
+    if len(consented_user_ids) != 0:
+        # This should pick up stragglers; messages we sent
+        # where we the recipient wasn't subscribed to by any of
+        # us (such as PMs to "them").
+        ids_of_non_exported_possible_recipients = ids_of_our_possible_senders - user_ids_for_us
+
+        recipients_for_them = Recipient.objects.filter(
+            type=Recipient.PERSONAL,
+            type_id__in=ids_of_non_exported_possible_recipients).values("id")
+        recipient_ids_for_them = get_ids(recipients_for_them)
+
+        messages_we_sent_to_them = Message.objects.filter(
+            sender__in=consented_user_ids,
+            recipient__in=recipient_ids_for_them,
+        ).order_by('id')
+
+        messages_we_received = Message.objects.filter(
+            Q(sender__in=consented_user_ids) | Q(recipient__in=recipient_ids_for_us),
+        ).exclude(recipient__in=recipient_ids_for_them).order_by('id')
+
+        message_queries = [
+            messages_we_received,
+            messages_we_sent_to_them,
+        ]
```

Comment:
I figured out a way to deduplicate this block with the "full export" code path, which I think makes things much more readable.